mater/lib/src/stores/file.rs

```
@@ -0,0 +1,306 @@
+        let count = index.len() as u64;
+        let entries = index
+            .into_iter()
+            .map(|(cid, offset)| IndexEntry::new(cid.hash().digest().to_vec(), offset as u64))
+            .collect();
+        let index = Index::MultihashIndexSorted(MultihashIndexSorted::from_single_width(
+            SHA_256_CODE,
+            SingleWidthIndex::new(Sha256::output_size() as u32, count, entries).into(),
+        ));
+        v2::write_index(&mut inner.store, &index).await?;
+
+        // Flush underlying writer
+        inner.store.flush().await?;
+        inner.is_finalized = true;
+
+        Ok(())
+    }
+}
+
+#[cfg(feature = "blockstore")]
+impl blockstore::Blockstore for FileBlockstore {
+    async fn get<const S: usize>(
+        &self,
+        cid: &CidGeneric<S>,
+    ) -> Result<Option<Vec<u8>>, blockstore::Error> {
+        let cid = Cid::try_from(cid.to_bytes()).map_err(|_err| blockstore::Error::CidTooLarge)?;
+
+        self.get(cid)
+            .await
+            .map_err(|err| blockstore::Error::FatalDatabaseError(err.to_string()))
+    }
+
+    async fn has<const S: usize>(&self, cid: &CidGeneric<S>) -> blockstore::Result<bool> {
+        let cid = Cid::try_from(cid.to_bytes()).map_err(|_err| blockstore::Error::CidTooLarge)?;
+
+        self.has(cid)
+            .await
+            .map_err(|err| blockstore::Error::FatalDatabaseError(err.to_string()))
+    }
+
+    async fn put_keyed<const S: usize>(
+        &self,
+        cid: &CidGeneric<S>,
+        data: &[u8],
+    ) -> Result<(), blockstore::Error> {
+        let cid = Cid::try_from(cid.to_bytes()).map_err(|_err| blockstore::Error::CidTooLarge)?;
+
+        self.put(&cid, data)
+            .await
+            .map_err(|err| blockstore::Error::FatalDatabaseError(err.to_string()))
+    }
+
+    async fn remove<const S: usize>(&self, _cid: &CidGeneric<S>) -> Result<(), blockstore::Error> {
+        unimplemented!("Operation not supported")
```

This panics, you should return an error instead